### PR TITLE
Refactor and Add Chaincode Unit Tests

### PR DESCRIPTION
**Pull request by btl5037**
Many chaincode developers raise questions on how to unit test chaincode. It used to be much easier with the old shim, as they could directly use the old mock stub in the shim. Now that it no longer exists the Asset Transfer example can provide an example of how to test Go chaincode using generated mocks.

Also worth noting is our current directory structure of our Go chaincodes prevents the creation of mocks due to import cycles. This change also pushes the chaincode implementation down into a `chaincode` package.

Signed-off-by: Brett Logan brett.t.logan@ibm.com

### CODE REVIEW

**Comment by sijocherian** [comment]
thanks for doing the changes suggested in #230 !

File: asset-transfer-basic/chaincode-go/chaincode/smartcontract.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+	contractapi.Contract
+}
+
+// Asset describes basic details of what makes up a simple asset
+type Asset struct {
+	ID             string `json:"ID"`
+	Color          string `json:"color"`
+	Size           int    `json:"size"`
+	Owner          string `json:"owner"`
+	AppraisedValue int    `json:"appraisedValue"`
```
Comment (sijocherian) [inline asset-transfer-basic/chaincode-go/chaincode/smartcontract.go:21]: suggest to follow snake_case for json keys. ``json:"appraised_value"``

**Comment by btl5037** [comment]
Will backfill with documentation after soliciting feedback

File: asset-transfer-basic/chaincode-go/chaincode/smartcontract.go (hunk 1 above)
Comment (btl5037) [reply asset-transfer-basic/chaincode-go/chaincode/smartcontract.go:21]: The example is generally idiomatic of the fabric standard core configuration. Go actually specifies that naming conventions should follow the conventions of the implementation where possible, and absent that they should at least follow a consistent convention. As JSON also specifies no standard convention, the fabric convention is used.

File: asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go [hunk 2]
```diff
@@ -0,0 +1,187 @@
+package chaincode_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"testing"
+
+	"github.com/hyperledger/fabric-protos-go/ledger/queryresult"
+
```
Comment (btl5037) [inline asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go:9]: Missed a blank line

File: asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+type stateQueryIterator interface {
+	shim.StateQueryIteratorInterface
+}
+
+func TestInitLedger(t *testing.T) {
+	chaincodeStub := &mocks.ChaincodeStub{}
+	transactionContext := &mocks.TransactionContext{}
+	transactionContext.GetStubReturns(chaincodeStub)
+
+	assetTransfer := chaincode.SmartContract{}
```
Comment (sijocherian) [inline asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go:36]: consider boilerplate code into a func

```
func setupMocks() (SmartContract, contractapi.TransactionContextInterface, *mocks.ChaincodeStub) {
	chaincodeStub := &mocks.ChaincodeStub{}
	transactionContext := &mocks.TransactionContext{}
	transactionContext.GetStubReturns(chaincodeStub)
	assetTransfer := SmartContract{}
	return assetTransfer, transactionContext, chaincodeStub
}
```

reuse like `assetTransfer, transactionContext, chaincodeStub := setupMocks()`

File: asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
+package chaincode_test
```
[the remark this answers is not in the document]
Comment (btl5037) [reply asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go:1]: `_test` is in an idiom in Go used to namespace the test. Unlike in other languages where it generally standard practice for tests to reside in a separate package entirely, the Go standard is that test files sit directly next to the test file with the `_test.go` suffix. In the event you need to perform tests as if they are in a totally separate package you can append `_test` to the package name an Go will compile it as a separate package, and soft-linking it to the main binary.

File: asset-transfer-basic/chaincode-go/chaincode/smartcontract.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+	}
+
+	return ctx.GetStub().PutState(id, assetJSON)
+}
+
+// ReadAsset returns the asset stored in the world state with given id.
+func (s *SmartContract) ReadAsset(ctx contractapi.TransactionContextInterface, id string) (*Asset, error) {
+	assetJSON, err := ctx.GetStub().GetState(id)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read from world state: %v", err)
```
[the remark this answers is not in the document]
Comment (btl5037) [reply asset-transfer-basic/chaincode-go/chaincode/smartcontract.go:85]: Correct, the entire `smartcontract.go` file was refactored into idiomatic and syntactically correct Go. The original in non-idiomatic in that, errors never start with a capital letter and and the `%v` placeholder is used to insert an error, you should never reach into the error and pull the string representation out, unless you need to parse it for some reason.

This entire file has been refactored into syntactically idiomatic Go.

File: asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+}
+
+func TestReadAsset(t *testing.T) {
+	chaincodeStub := &mocks.ChaincodeStub{}
+	transactionContext := &mocks.TransactionContext{}
+	transactionContext.GetStubReturns(chaincodeStub)
+
+	expectedAsset := &chaincode.Asset{ID: "asset1"}
+	bytes, err := json.Marshal(expectedAsset)
+	require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (btl5037) [reply asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go:70]: We never ignore errors. There is absolutely no reason not to check it here.

File: asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+	chaincodeStub.GetStateReturns(bytes, nil)
+	assetTransfer := chaincode.SmartContract{}
+	asset, err := assetTransfer.ReadAsset(transactionContext, "")
```
[the remark this answers is not in the document]
Comment (btl5037) [reply asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go:74]: `asset1` has no influence on this test, and thus the zero value is used to make it clear that there is no dependency on the input of the parameter to the outcome of the test.

File: asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+	transactionContext := &mocks.TransactionContext{}
+	transactionContext.GetStubReturns(chaincodeStub)
+
+	asset := &chaincode.Asset{ID: "asset1"}
+	bytes, err := json.Marshal(asset)
+	require.NoError(t, err)
+
+	chaincodeStub.GetStateReturns(bytes, nil)
+	assetTransfer := chaincode.SmartContract{}
+	err = assetTransfer.TransferAsset(transactionContext, "", "")
```
[the remark this answers is not in the document]
Comment (btl5037) [reply asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go:146]: `asset1` and `Jerry` have no influence on this test, and thus the zero value is used to make it clear that there is no dependency on the input of the parameter to the outcome of the test.

File: asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sijocherian) [reply asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go:1]: hmm, interesting

File: asset-transfer-basic/chaincode-go/chaincode/smartcontract.go [hunk 9]
```diff
@@ -0,0 +1,193 @@
...
+			return fmt.Errorf("failed to put to world state. %v", err)
+		}
+	}
+
+	return nil
+}
+
+// CreateAsset issues a new asset to the world state with given details.
+func (s *SmartContract) CreateAsset(ctx contractapi.TransactionContextInterface, id, color, owner string, size, appraisedValue int) error {
+	exists, err := s.AssetExists(ctx, id)
```
Comment (sijocherian) [inline asset-transfer-basic/chaincode-go/chaincode/smartcontract.go:58]: more of input validation in this chaincode, on `id`, will be a good example for developers:  CreateAsset, or possibly AssetExists , should check if `id` is Empty and throw error, to disallow Assets with empty id.

Comment (denyeart) [reply asset-transfer-basic/chaincode-go/chaincode/smartcontract.go:58]: Good idea, let's do in a subsequent PR. Feel free to contribute it.

File: asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go (hunk 3 above)
Comment (denyeart) [reply asset-transfer-basic/chaincode-go/chaincode/smartcontract_test.go:36]: Good idea, feel free to contribute in subsequent PR.

**Pull request merged**